_posts/2019-03-18-rules-jvm-external-maven.md

```diff
@@ -0,0 +1,159 @@
+particular, this means that users only need to specify their direct dependencies
+and the rule will handle the rest for them.
+
+## Features and use cases
+
+For the initial release, we are targeting support for Java and Android Java
+projects. The following features are supported:
+
+* Supported on Windows, macOS and Linux
+* JAR and AAR packaging types
+* In-band transitive closure resolution
+* Dependency specification directly in the WORKSPACE file
+* Custom Maven repositories
+* Maven repository authentication with HTTP Basic Authentication
+* Artifact version pinning and resolution
+* Multiple independent closures of versioned artifacts
+    * Not forced to only have one version of an artifact per workspace, although
+      different versions are namespaced by the repository name
+* Source JAR support
+* Dependency exclusion
+
+As part of this, we will no longer be maintaining generate_workspace.
+gmaven_rules, which was designed to be an interim solution, will also be
+replaced by rules_jvm_external. We will be updating existing documentation to
+reflect this over the coming weeks.
+
+## How to use
+
+To start using this project in your Java or Android project, in your WORKSPACE
+file, point to the repository rule:
+
+```python
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+http_archive(
+    name = "rules_jvm_external",
+    strip_prefix = "rules_jvm_external-1.1",
+    sha256 = "ade316ec98ba0769bb1189b345d9877de99dd1b1e82f5a649d6ccbcb8da51c1f",
+    url = "https://github.com/bazelbuild/rules_jvm_external/archive/1.1.zip"
+)
+```
+
+Then the key components of the rule are to specify your direct dependencies
+using the `artifacts` attribute and where those artifacts should come from using
+the `repositories` attribute:
+
+```python
+load("@rules_jvm_external//:defs.bzl", "maven_install")
+
+maven_install(
+    name = "maven",
+    artifacts = [
+        "com.google.guava:guava:27.0-android",
+        "androidx.test.espresso:espresso-core:3.1.1",
+    ],
+    repositories = [
+        "https://maven.google.com",
+        "https://repo1.maven.org/maven2",
+    ],
+    fetch_sources = True,   # Fetch source jars. Defaults to False.
+)
+```
+
+The `maven_install` rule creates targets you can then depend on in your `BUILD`
+files. For example:
+
+```python
+android_library(
+    name = "foo",
+    srcs = ["MyAndroidLib.java"],
+    deps = [
+        "@maven//:androidx_test_espresso_espresso_core"
+        "@maven//:com_google_guava_guava",
+    ],
+)
+```
+
+We have also added a `artifact` helper function to specify the Maven coordinates
+directly. For example, the previous example can be written as:
+
+```python
+
+load("@rules_jvm_external//:defs.bzl", "artifact")
+
+android_library(
+    name = "foo",
+    srcs = ["MyAndroidLib.java"],
+    deps = [
+        artifact("androidx.test.espresso:espresso-core"),
+        artifact("com.google.guava:guava"),
```

Comment:
Looking at both examples, it's not clear to me if the `artifact` function is actually useful.
(the abstraction can confuse readers of the file, as well as tools manipulating BUILD files)

Reply:
It's meant to guide users coming from Gradle into a familiar syntax. We have previously used this approach in gmaven_rules to large success, because the generated target names were just way too long and tedious to type out manually.

I don't think limitations by tools manipulating BUILD files should limit expressiveness in them.

Reply:
On further thought, I think we can cut this snippet out of the blogpost, since it's already documented. @dkelmer?

Reply:
Removed to make the blogpost simpler.

Reply:
Hey I like the `artifact` function (especially if it makes things more familiar to users)

Reply:
@laurentlb is that ok with you?

Reply:
I'm with @laurentlb on this one. I don't think that 
`artifact("com.google.guava:guava")` is more readable than `@maven//com_google_guava_guava`. I think we need to keep supporting it because gmaven_rules had it, but I'm not sure it should be recommended.

Reply:
Summarized from offline conversation with Jin:

General argument for the "bazel-y" syntax: This is a bazel project, using bazel syntax makes sense (especially since the distinction between the two doesn't actually make the dependency incomprehensible even if you only have maven experience).
=============

Argument for keeping the artifact macro: It's nice that the workspace file and the build files have the same syntax to specify the dependencies, especially since you can copy paste. 
Response to this argument: It's ok that the workspace file and the build files have different syntax. After all, workspace files interact with the "outside world" and other systems so it makes sense that the syntax in the workspace file aligns with the syntax those other systems expect. But once you're in build file land, that's strictly bazel land and so it makes sense to use bazel-y syntax. 
=============

Reply:
@dkelmer's point about the split between BUILD and WORKSPACE spaces is convincing to me. The interface between Bazel and Maven is `maven_install`'s API in the WORKSPACE file, and we shouldn't recommend an "escape hatch" abstraction in BUILD files, even though it's nice to have.

Reply:
Either is fine with me (I don't have a strong opinion), but I'd suggest to keep only one example, the one you want to recommend. 

> For example, the previous example can be written

I'm afraid half the users will pick one example and the other half will pick the other. If the `artifact` function provides a better experience, don't show what it expands too (it's an implementation detail).

Reply:
> If the artifact function provides a better experience, don't show what it expands too (it's an implementation detail).

It does, only on a surface level. Strict Java deps will recommend the actual label to be added into the BUILD file, and not the artifact macro. The strict deps error appears pretty often too, if users depend on classes from the transitive classpath.

With all these reasons, I'll remove it.

Reply:
Thanks for the great discussion, I am fine with this.